File: 496. Next Greater Element I.cpp
class Solution {
public:
    // brute force: O(N2)time, O(N)space, 67.69%
    vector<int> nextGreaterElement(vector<int>& findNums, vector<int>& nums) {
        vector<int> ret;
        if(findNums.empty() || nums.empty())
            return ret;
        unordered_map<int, int> greater;
        for(int i = 0; i < nums.size(); i++){
            greater[nums[i]] = -1;      // initialize
            for(int j = i + 1; j < nums.size(); j++){
                if(nums[j] > nums[i]){
                    greater[nums[i]] = nums[j];
                    break;
                }
            }
        }
        for(int i = 0; i < findNums.size(); i++)
            ret.push_back(greater[findNums[i]]);
        return ret;
    }

    // dp(backwards scan): O(N2)time, O(N)space, 67.69%
    vector<int> nextGreaterElement(vector<int>& findNums, vector<int>& nums) {
        vector<int> ret;
        if(findNums.empty() || nums.empty())
            return ret;
        unordered_map<int, int> greater = {{nums.back(), -1}};
        for(int i = nums.size() - 2; i >= 0; i--){
            greater[nums[i]] = -1;  // initialize
            if(nums[i] < nums[i + 1])
                greater[nums[i]] = nums[i + 1];
            else{
                for(int j = i + 1; j < nums.size(); j++){
                    if(nums[i] < greater[nums[j]]){
                        greater[nums[i]] = greater[nums[j]];
                        break;
                    }
                }
            }
        }
        for(int i = 0; i < findNums.size(); i++)
            ret.push_back(greater[findNums[i]]);
        return ret;
    }

    // stack storing decreasing seq: O(N)time & space, 67.69%, ok
    vector<int> nextGreaterElement(vector<int>& findNums, vector<int>& nums) {
        vector<int> ret;
        if(findNums.empty() || nums.empty())
            return ret;
        stack<int> desc;  // stack to store descending sequence
        unordered_map<int, int> greater;
        for(auto n: nums){
            while(desc.size() && desc.top() < n){
                greater[desc.top()] = n;
                desc.pop();
            }
            desc.push(n);
        }
        for(int i = 0; i < findNums.size(); i++)
            ret.push_back(greater.count(findNums[i])? greater[findNums[i]]: -1);
        return ret;
    }
};s

File: 496. Next Greater Element I.py
class Solution(object):
    
    # brute force: O(mn) time, O(1) space
    def nextGreaterElement(self, findNums, nums):
        """
        :type findNums: List[int]
        :type nums: List[int]
        :rtype: List[int]
        """
        def findGreater(nums, target):
            found = False
            for num in nums:
                if num == target:
                    found = True
                if found and target < num:
                    return num
            return -1
        ret = []
        for num in findNums:
            ret.append(findGreater(nums, num))
        return ret
    
    # hash table: O(n2) time, O(n) space
    def nextGreaterElement(self, findNums, nums):
        greater = {num: -1 for num in nums}
        for i in xrange(len(nums) - 1):
            for j in xrange(i + 1, len(nums)):
                if nums[j] > nums[i]:
                    greater[nums[i]] = nums[j]
                    break
        return [greater[num] for num in findNums]
    
    # !! stack & hashtable: O(n+m) time, O(n) space
    def nextGreaterElement(self, findNums, nums):
        stack, greater = [], {}
        for num in nums:
            while len(stack) and stack[-1] < num:
                greater[stack.pop()] = num
            stack.append(num)
        while stack:
            greater[stack.pop()] = -1
        return [greater[num] for num in findNums]
                
        